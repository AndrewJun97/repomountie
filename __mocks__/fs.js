// __mocks__/fs.js
'use strict';

const fs = jest.requireActual('fs');

function access(path, flag, cb) {
  cb(undefined);
}

function readFile(path, options, cb) {
  cb(undefined, Buffer.from('Hello World', 'utf8'));
  // cb(undefined, 'Hello World');
}

fs.readFile = readFile;
fs.access = access;

module.exports = fs;
