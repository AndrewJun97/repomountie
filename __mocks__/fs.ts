// __mocks__/fs.js
'use strict';

const fs = require.requireActual('fs');

function access(path, flag, cb) {
  if (path === 'no-file-access') {
    return cb(new Error('No access to this file - mock'));
  }

  return cb(undefined);
}

function readFile(path, options, cb) {
  if (path === 'no-file') {
    return cb(new Error('No such file - mock'), undefined);
  }
  const text = 'Hello World';
  return cb(undefined, typeof options === 'undefined' ? Buffer.from(text, 'utf8') : text);
}

fs.readFile = readFile;
fs.access = access;

module.exports = fs;
