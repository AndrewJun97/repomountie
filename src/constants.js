"use strict";
//
// Copyright Â© 2018, 2019, 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2018-10-01.
//
exports.__esModule = true;
exports.SCHEDULER_DELAY = 24 * 60 * 60 * 1000; // one day
exports.COMMENT_TRIGGER_WORD = 'help';
exports.GITHUB_ID = 'repo-mountie';
exports.REPO_CONFIG_FILE = 'rmconfig.json';
exports.HELP_DESK = {
    SUPPORT_USERS: ['jleach']
};
exports.COMMIT_FILE_NAMES = {
    COMPLIANCE: 'COMPLIANCE.yaml',
    LICENSE: 'LICENSE'
};
exports.COMMIT_MESSAGES = {
    ADD_COMPLIANCE: 'Add compliance audit file',
    ADD_LICENSE: 'Add Apache License 2.0',
    UPDATE_COMPLIANCE: 'Updating compliance audit file'
};
exports.PR_TITLES = {
    ADD_COMPLIANCE: 'Add missing compliance audit file',
    ADD_LICENSE: 'Add missing license'
};
exports.BRANCHES = {
    ADD_COMPLIANCE: 'repo-mountie/add-compliance',
    ADD_LICENSE: 'repo-mountie/add-license'
};
exports.TEMPLATES = {
    COMPLIANCE: 'templates/COMPLIANCE.yaml',
    CONDUCT: 'templates/CODE_OF_CONDUCT.md',
    CONTRIBUTE: 'templates/CONTRIBUTING.md',
    LICENSE: 'templates/LICENSE',
    README: 'templates/README.md'
};
exports.VALID_LICENSES = {
    APACHE: 'apache-2.0'
};
exports.TEXT_FILES = {
    HOWTO_PR: 'templates/howto_pull_request.md',
    STALE_COMMENT: 'templates/stale_issue_comment.md',
    WHY_COMPLY: 'templates/why-comply.md',
    WHY_LICENSE: 'templates/why-license.md'
};
exports.COMMANDS = {
    IGNORE: '/bot-ignore-length'
};
exports.REGEXP = {
    command: '@repo-mountie\\s+',
    compliance: '@repo-mountie\\s+update-(pia|stra)\\s+(in-progress|completed|TBD|exempt)',
    help: '@repo-mountie\\s+help'
};
exports.ACCESS_CONTROL = {
    allowedInstallations: ['bcgov', 'fullboar'],
    allowedSsoClients: ['reggie-api']
};
