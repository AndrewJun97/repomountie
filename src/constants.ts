//
// Repo Mountie
//
// Copyright Â© 2018 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2018-10-01.
//

export const SCHEDULER_DELAY: number = 24 * 60 * 60 * 1000; // one day

export const HELP_DESK = {
  LICENSE_SUPPORT_USERS: ['jleach'],
};

export const PR_TITLES = {
  ADD_LICENSE: 'Add missing license',
};

export const BRANCHES = {
  ADD_LICENSE: 'fix/add-license',
};

export const TEMPLATES = {
  CONDUCT: 'templates/CODE_OF_CONDUCT.md',
  CONTRIBUTE: 'templates/CONTRIBUTING.md',
  LICENSE: 'templates/LICENSE',
  README: 'templates/README.md',
};

export const VALID_LICENSES = {
  APACHE: 'apache-2.0',
};

export const TEXT_FILES = {
  WHY_LICENSE: 'templates/why-license.md',
};

export const ALLOWED_REPOS = [
  'repo-mountie',
  'devhub-signing-api',
  'devhub-signing-web',
  'devhub-signing-agent',
  'range-api',
  'range-web',
];
