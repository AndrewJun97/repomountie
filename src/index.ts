//
// Copyright Â© 2018, 2019, 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2018-10-01.
//

import { logger } from '@bcgov/common-nodejs-utils';
import mongoose from 'mongoose';
import { Application, Context } from 'probot';
import createScheduler from 'probot-scheduler';
import config from './config';
import { ACCESS_CONTROL, SCHEDULER_DELAY } from './constants';
import { fetchComplianceFile, fetchConfigFile } from './libs/ghutils';
import { checkForStaleIssues, created } from './libs/issue';
import { validatePullRequestIfRequired } from './libs/pullrequest';
import { addLicenseIfRequired, addSecurityComplianceInfoIfRequired } from './libs/repository';
import { routes } from './libs/routes';
import { extractComplianceStatus } from './libs/utils';

process.env.TZ = 'UTC';

if (['development', 'test'].includes(process.env.NODE_ENV || 'development')) {
  process.on('unhandledRejection', (reason, p) => {
    // @ts-ignore: `stack` does not exist on type
    const stack = typeof (reason) !== 'undefined' ? reason.stack : 'unknown';
    // Token decode errors are OK in test because we use a faux token for
    // mock objects.
    if (stack.includes('HttpError: A JSON web token could not be decoded')) {
      return;
    }

    logger.warn(`Unhandled rejection at promise = ${JSON.stringify(p)}, reason = ${stack}`);
  });
}

export = async (app: Application) => {
  logger.info('Robot Loaded!!!');

  routes(app);

  const scheduler = createScheduler(app, {
    delay: false, // !!process.env.DISABLE_DELAY, // delay is enabled on first run
    interval: SCHEDULER_DELAY,
  });

  app.on('pull_request.opened', pullRequestOpened);
  app.on('issue_comment.created', issueCommentCreated);
  app.on('schedule.repository', repositoryScheduled);
  app.on('repository.deleted', repositoryDelete);
  // app.on('repository_vulnerability_alert.create', blarb);

  try {
    const options = {};
    const user = config.get('db:user');
    const passwd = config.get('db:password');
    const host = config.get('db:host');
    const dbname = config.get('db:database');
    const curl = `mongodb://${user}:${passwd}@${host}/${dbname}`;

    await mongoose.connect(curl, options);
  } catch (err) {
    const message = `Unable to open database connection`;
    throw new Error(`${message}, error = ${err.message}`);
  }

  async function pullRequestOpened(context: Context) {
    try {
      const owner = context.payload.installation.account.login;
      const isFromBot = context.isBot;

      if (!ACCESS_CONTROL.allowedInstallations.includes(owner)) {
        logger.info(
          `Skipping PR ${context.payload.pull_request.number} for repo ${
          context.payload.repository.name
          } because its not from an allowed installation`
        );
        return;
      }

      // This can throw a `TypeError` during testing.
      if (isFromBot) {
        // Don't act crazy.
        logger.info(
          `Skipping PR ${context.payload.pull_request.number} for repo ${
          context.payload.repository.name
          } because its from a bot`
        );
        return;
      }
    } catch (err) {
      logger.info(`Unable to handle pull request, err = ${err.message}`);
    }

    logger.info(
      `Processing PR ${context.payload.pull_request.number} for repo ${
      context.payload.repository.name
      }`
    );

    const rmconfig = await fetchConfigFile(context);

    await validatePullRequestIfRequired(context, rmconfig);
  }

  async function issueCommentCreated(context: Context) {
    try {
      const owner = context.payload.organization.login;
      const isFromBot = context.isBot;

      if (!ACCESS_CONTROL.allowedInstallations.includes(owner)) {
        logger.info(
          `Skipping issue ${context.payload.pull_request.number} for repo ${
          context.payload.repository.name
          } because its not from an allowed installation`
        );
        return;
      }

      // This can throw a `TypeError` during testing.
      if (isFromBot) {
        // Don't act crazy.
        logger.info(`Skipping issue ${context.payload.issue.id} because its from a bot`);
        return;
      }
    } catch (err) {
      logger.info(`Unable to process issue comment, err = ${err.message}`);
    }

    logger.info(`Processing issue ${context.payload.issue.id}`);

    try {
      await created(context);
    } catch (err) {
      logger.error(`Unable to process issue ${context.payload.issue.id}`);
    }
  }

  async function repositoryDelete(context: Context) {
    scheduler.stop(context.payload.repository);
  }

  async function repositoryScheduled(context: Context) {
    logger.info(`Processing ${context.payload.repository.name}`);

    const owner = context.payload.installation.account.login;
    if (!ACCESS_CONTROL.allowedInstallations.includes(owner)) {
      logger.info(
        `Skipping scheduled repository ${
        context.payload.repository.name
        } because its not part of an allowed installation`
      );
      return;
    }

    if (context.payload.repository.archived) {
      logger.warn(`The repo ${context.payload.repository.name} is archived. Skipping.`);
      return;
    }

    let requiresComplianceFile = false;
    try {
      const data = await fetchComplianceFile(context);
      const doc = extractComplianceStatus(context.payload.repository.name,
        context.payload.installation.account.login, data);

      await doc.save();
    } catch (err) {
      const message = `Unable to check compliance in repository ${context.payload.repository.name}`;
      logger.error(`${message}, error = ${err.message}`);

      requiresComplianceFile = true;
    }

    try {
      if (requiresComplianceFile) {
        await addSecurityComplianceInfoIfRequired(context, scheduler);
      }
      await addLicenseIfRequired(context, scheduler);

      // Functionality below here requires a `config` file exist in the repo.

      try {
        const rmconfig = await fetchConfigFile(context);
        await checkForStaleIssues(context, rmconfig);
      } catch (err) {
        logger.info('No config file. Skipping.');
      }
    } catch (err) {
      const message = `Unable to process repository ${context.payload.repository.name}`;
      logger.error(`${message}, error = ${err.message}`);
    }
  }
};
