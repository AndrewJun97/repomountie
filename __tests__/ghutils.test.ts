//
// Copyright Â© 2018, 2019 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2018-10-04.
//

import fs from 'fs';
import path from 'path';
import { Context } from 'probot';
import { COMMIT_FILE_NAMES, ISSUE_TITLES } from '../src/constants';
import { addFileViaPullRequest, assignUsersToIssue, checkIfFileExists, checkIfRefExists, fetchCollaborators, fetchComplianceFile, fetchConfigFile, fetchContentsForFile, fetchFileContent, fetchPullRequests, hasPullRequestWithTitle, isOrgMember, labelExists } from '../src/libs/ghutils';
import helper from './src/helper';

const p0 = path.join(__dirname, 'fixtures/repo-schedule-event.json');
const repoScheduledEvent = JSON.parse(fs.readFileSync(p0, 'utf8'));

const p1 = path.join(__dirname, 'fixtures/repo-get-content-compliance.json');
const complianceResponse = JSON.parse(fs.readFileSync(p1, 'utf8'));

const p2 = path.join(__dirname, 'fixtures/repo-get-content-config.json');
const configResponse = JSON.parse(fs.readFileSync(p2, 'utf8'));

const p3 = path.join(__dirname, 'fixtures/master.json');
const master = JSON.parse(fs.readFileSync(p3, 'utf8'));

const p5 = path.join(__dirname, 'fixtures/repo-get-pulls.json');
const listPulls = JSON.parse(fs.readFileSync(p5, 'utf8'));

const p6 = path.join(__dirname, 'fixtures/repo-list-commits.json');
const listCommits = JSON.parse(fs.readFileSync(p6, 'utf8'));

const p7 = path.join(__dirname, 'fixtures/org-user-ismember.json');
const memberhip = JSON.parse(fs.readFileSync(p7, 'utf8'));

const p8 = path.join(__dirname, 'fixtures/repo-get-collaborators-response.json');
const collabs = JSON.parse(fs.readFileSync(p8, 'utf8'));

describe('GitHub utility functions', () => {
    let context;
    const { github } = helper;

    beforeEach(() => {
        context = new Context(repoScheduledEvent, github as any, {} as any);
        context.payload.organization = {
            login: 'bcgov',
        };
    });

    afterEach(() => {
        jest.clearAllMocks();
        jest.resetAllMocks();
    });

    it('Labels should be fetched for lookup', async () => {
        await labelExists(context, 'blarb');
        expect(github.issues.listLabelsForRepo).toHaveBeenCalled();
    });

    it('A file should be retrieved.', async () => {
        github.repos.getContents = jest.fn().mockReturnValueOnce(Promise.resolve(complianceResponse));
        const data = await fetchFileContent(context, COMMIT_FILE_NAMES.COMPLIANCE);

        expect(data).toMatchSnapshot();
    });

    it('A file should not be retrieved.', async () => {
        await expect(fetchFileContent(context, 'blarb.txt')).rejects.toThrow(Error);
    });

    it('The compliance file should be retrieved.', async () => {
        github.repos.getContents = jest.fn().mockReturnValueOnce(Promise.resolve(complianceResponse));
        const data = await fetchComplianceFile(context);

        expect(github.repos.getContents).toHaveBeenCalled();
        expect(data).toMatchSnapshot();
    });

    it('The config file should be retrieved.', async () => {
        github.repos.getContents = jest.fn().mockReturnValueOnce(Promise.resolve(configResponse));
        const data = await fetchConfigFile(context);

        expect(github.repos.getContents).toHaveBeenCalled();
        expect(data).toMatchSnapshot();
    });

    it('The ref should exists.', async () => {
        github.git.getRef.mockReturnValueOnce(master);
        const result = await checkIfRefExists(context, context.payload.repository.default_branch);

        expect(github.git.getRef).toHaveBeenCalled();
        expect(result).toBeTruthy();
    });

    it('The ref should not exists.', async () => {
        github.git.getRef.mockReturnValueOnce(Promise.reject());
        const result = await checkIfRefExists(context, context.payload.repository.default_branch);

        expect(github.git.getRef).toHaveBeenCalled();
        expect(result).toBeFalsy();
    });

    it('A file should be added by PR', async () => {
        const owner = 'bcgov';
        const repo = 'hello5';
        github.git.getRef.mockReturnValueOnce(master);

        await addFileViaPullRequest(context, owner, repo, 'Hello', 'World',
            'This is a body.', 'blarb', 'blarb.txt', 'some-data-here');

        expect(github.git.getRef).toHaveBeenCalled();
        expect(github.pulls.create).toHaveBeenCalled();
        expect(github.git.createRef).toHaveBeenCalled();
        expect(github.repos.createOrUpdateFile).toHaveBeenCalled();
    });

    it('Pull requests are retrieved', async () => {
        github.pulls.list = jest.fn().mockReturnValueOnce(Promise.resolve(listPulls));
        const results = await fetchPullRequests(context);

        expect(results).toMatchSnapshot();
    });

    it('A pull request should not exists', async () => {
        github.pulls.list = jest.fn().mockReturnValueOnce(Promise.resolve(listPulls));
        const result = await hasPullRequestWithTitle(context, 'Hello');

        expect(result).toBeFalsy();
    });

    it('A pull request should exists', async () => {
        github.pulls.list = jest.fn().mockReturnValueOnce(Promise.resolve(listPulls));
        const result = await hasPullRequestWithTitle(context, ISSUE_TITLES.ADD_LICENSE);

        expect(result).toBeTruthy();
    });

    it('Assigning an issue on GitHub should succeed', async () => {
        github.issues.addAssignees = jest.fn().mockReturnValueOnce(Promise.resolve());

        await expect(assignUsersToIssue(context, ['blarb'])).resolves.toBeUndefined();
    });

    it('Assigning an issue on GitHub should fail', async () => {
        github.issues.addAssignees = jest.fn().mockReturnValueOnce(Promise.reject());

        await expect(assignUsersToIssue(context, ['blarb'])).rejects.toThrow();
    });

    it('File contents should be retrieved', async () => {
        github.repos.getContents = jest.fn().mockReturnValueOnce(Promise.resolve(complianceResponse));
        github.repos.listCommits = jest.fn().mockReturnValueOnce(Promise.resolve(listCommits));

        const results = await fetchContentsForFile(context, 'helo.yaml');

        expect(github.repos.listCommits).toHaveBeenCalled();
        expect(github.repos.getContents).toHaveBeenCalled();
        expect(results).toMatchSnapshot();
    });

    it('A user is a member of the organization', async () => {
        github.orgs.checkMembership = jest.fn().mockReturnValueOnce(Promise.resolve(memberhip));

        const result = await isOrgMember(context, 'helloworld');
        expect(result).toBeTruthy();
    });

    it('A user is not a member of the organization', async () => {
        const err: any = new Error('Nope');
        err.code = 404;

        github.orgs.checkMembership = jest.fn().mockReturnValueOnce(Promise.reject(err));

        const result = await isOrgMember(context, 'helloworld');
        expect(result).toBeFalsy();
    });

    it('A user lookup fails unexpectedly', async () => {
        const err: any = new Error('Nope');

        github.orgs.checkMembership = jest.fn().mockReturnValueOnce(Promise.reject(err));

        await expect(isOrgMember(context, 'helloworld')).rejects.toThrow();
    });

    it('A user lookup fails due to an unexpected http code', async () => {
        memberhip.status = 512;
        github.orgs.checkMembership = jest.fn().mockReturnValueOnce(Promise.resolve(memberhip));

        const result = await isOrgMember(context, 'helloworld');
        expect(result).toBeFalsy();
    });

    it('Return true if a file exists', async () => {
        github.repos.getContents = jest.fn().mockReturnValueOnce(Promise.resolve(complianceResponse));

        await expect(checkIfFileExists(context, 'hello.yaml')).resolves.toBeTruthy();
    });

    it('Return false if a file does not exist', async () => {
        github.repos.getContents = jest.fn().mockReturnValueOnce(Promise.reject());

        await expect(checkIfFileExists(context, 'hello.yaml')).resolves.toBeFalsy();
    });

    it('Fetching collaborators should succeed', async () => {
        github.repos.listCollaborators = jest.fn().mockReturnValueOnce(Promise.resolve(collabs));

        const results = await fetchCollaborators(context);

        expect(results).toMatchSnapshot();
    });
});
