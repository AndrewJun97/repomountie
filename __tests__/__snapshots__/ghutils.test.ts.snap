// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitHub utility functions A file should be retrieved. 1`] = `
"---
name: compliance
description: |
  This document is used to track a projects PIA and STRA
  compliance.
spec:
  - name: PIA
    status: completed
    last-updated: \\"2019-11-22T00:03:52.138Z\\"
  - name: STRA
    status: in-progress
    last-updated: \\"2019-11-22T00:03:52.138Z\\"
"
`;

exports[`GitHub utility functions File contents should be retrieved 1`] = `
Object {
  "_links": Object {
    "git": "https://api.github.com/repos/bcgov/glarb/git/blobs/fc5f44d45e03b298219083b201db858f7a06aa1d",
    "html": "https://github.com/bcgov/glarb/blob/master/compliance.yaml",
    "self": "https://api.github.com/repos/bcgov/glarb/contents/compliance.yaml?ref=master",
  },
  "content": "LS0tCm5hbWU6IGNvbXBsaWFuY2UKZGVzY3JpcHRpb246IHwKICBUaGlzIGRvY3VtZW50IGlzIHVzZWQgdG8gdHJhY2sgYSBwcm9qZWN0cyBQSUEgYW5kIFNUUkEKICBjb21wbGlhbmNlLgpzcGVjOgogIC0gbmFtZTogUElBCiAgICBzdGF0dXM6IGNvbXBsZXRlZAogICAgbGFzdC11cGRhdGVkOiAiMjAxOS0xMS0yMlQwMDowMzo1Mi4xMzhaIgogIC0gbmFtZTogU1RSQQogICAgc3RhdHVzOiBpbi1wcm9ncmVzcwogICAgbGFzdC11cGRhdGVkOiAiMjAxOS0xMS0yMlQwMDowMzo1Mi4xMzhaIgo=
",
  "download_url": "https://raw.githubusercontent.com/bcgov/glarb/master/compliance.yaml",
  "encoding": "base64",
  "git_url": "https://api.github.com/repos/bcgov/glarb/git/blobs/fc5f44d45e03b298219083b201db858f7a06aa1d",
  "html_url": "https://github.com/bcgov/glarb/blob/master/compliance.yaml",
  "name": "compliance.yaml",
  "path": "compliance.yaml",
  "sha": "fc5f44d45e03b298219083b201db858f7a06aa1d",
  "size": 50,
  "type": "file",
  "url": "https://api.github.com/repos/bcgov/glarb/contents/compliance.yaml?ref=master",
}
`;

exports[`GitHub utility functions The compliance file should be retrieved. 1`] = `
Object {
  "description": "This document is used to track a projects PIA and STRA
compliance.
",
  "name": "compliance",
  "spec": Array [
    Object {
      "last-updated": "2019-11-22T00:03:52.138Z",
      "name": "PIA",
      "status": "completed",
    },
    Object {
      "last-updated": "2019-11-22T00:03:52.138Z",
      "name": "STRA",
      "status": "in-progress",
    },
  ],
}
`;

exports[`GitHub utility functions The config file should be retrieved. 1`] = `
Object {
  "pullRequest": Object {
    "maxLinesChanged": 300,
  },
}
`;
